## 6-е домашнее задание - Библиотека numpy

<b>Срок до 2 июня 2022г. (включительно)</b>

Подпишите этот блокнот: укажите ФИО и номер группы. Выполненное задание загрузите в свой репозиторий «<b>Spring_Python_2022</b>» в github в папку "<b>homework-6</b>".  

Перед выполнением этого домашнего задания полезно прочитать про возможности и основные методы библиотеки numpy:  
http://acm.mipt.ru/twiki/bin/view/Cintro/PythonNumpy  


Кондрашова Анастасия
НПМбд-01-19

In [1]:
import numpy as np
import random

1.	Создайте равномерное разбиение интервала от -1.3 до 2.5 на 64 отрезка.

In [2]:
np.linspace(-1.3, 2.5, 64)

array([-1.3       , -1.23968254, -1.17936508, -1.11904762, -1.05873016,
       -0.9984127 , -0.93809524, -0.87777778, -0.81746032, -0.75714286,
       -0.6968254 , -0.63650794, -0.57619048, -0.51587302, -0.45555556,
       -0.3952381 , -0.33492063, -0.27460317, -0.21428571, -0.15396825,
       -0.09365079, -0.03333333,  0.02698413,  0.08730159,  0.14761905,
        0.20793651,  0.26825397,  0.32857143,  0.38888889,  0.44920635,
        0.50952381,  0.56984127,  0.63015873,  0.69047619,  0.75079365,
        0.81111111,  0.87142857,  0.93174603,  0.99206349,  1.05238095,
        1.11269841,  1.17301587,  1.23333333,  1.29365079,  1.35396825,
        1.41428571,  1.47460317,  1.53492063,  1.5952381 ,  1.65555556,
        1.71587302,  1.77619048,  1.83650794,  1.8968254 ,  1.95714286,
        2.01746032,  2.07777778,  2.13809524,  2.1984127 ,  2.25873016,
        2.31904762,  2.37936508,  2.43968254,  2.5       ])

2.	Сгенерируйте numpy массив длины $3n$, заполненный циклически числами 1, 2, 3, 1, 2, 3, 1....

In [3]:
n = 5
np.arange(3*n) % 3 + 1

array([1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3])

3.	Создайте массив первых 10 нечетных целых чисел.

In [4]:
np.arange(1, 20, 2)

array([ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19])

4.  Создайте массив нулей размера 10 x 10, а затем создайте в нём "рамку" из единиц по краям.

In [5]:
Matrix = np.zeros((10, 10))
Matrix[::9,:] = 1
Matrix[1:9,::9] = 1
Matrix

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

5.  Создайте массив 8 x 8 с шахматной доской из нулей и единиц.

In [8]:
Matrix = np.zeros((8, 8))
Matrix[1::2,1::2] = 1
Matrix[::2,::2] = 1
Matrix

array([[1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.]])

6.  Создайте функцию, которая создает $n×n$ матрицу с $(i,j)$-элементом, равным $i+j$.

In [10]:
def create_matrix(n):
    A = np.arange(n**2)
    return (A // n + A % n).reshape(n, n)

create_matrix(4)

array([[0, 1, 2, 3],
       [1, 2, 3, 4],
       [2, 3, 4, 5],
       [3, 4, 5, 6]])

7. Примените функцию dot для перемножения вектор на вектор, матрицу на вектор и матрицу на матрицу.

In [17]:
vector1 = np.arange(1, 2)
vector2 = np.arange(3, 4)
print('vector1 x vector2 =', np.dot(vector1, vector2))

matrix1 = np.arange(1, 7).reshape((2, 3))
matrix2 = np.arange(1, 7).reshape((3, 2))
print('matrix1 x matrix2 =\n', np.dot(matrix1, matrix2))

vector1 x vector2 = 3
matrix1 x matrix2 =
 [[22 28]
 [49 64]]


8. Вычислите cos(x) и sin(x) на интервале [0,1] с шагом $0.05$, а затем объедините оба массива чисел как строки в один массив.

In [20]:
interval = np.linspace(0, 1, int(1/0.05)+1)
Cosinus = np.cos(interval)
Sinus = np.sin(interval)
print('sin =', Cosinus, '\ncos =', Sinus)
np.hstack([Cosinus, Sinus])

sin = [1.         0.99875026 0.99500417 0.98877108 0.98006658 0.96891242
 0.95533649 0.93937271 0.92106099 0.9004471  0.87758256 0.85252452
 0.82533561 0.7960838  0.76484219 0.73168887 0.69670671 0.65998315
 0.62160997 0.58168309 0.54030231] 
cos = [0.         0.04997917 0.09983342 0.14943813 0.19866933 0.24740396
 0.29552021 0.34289781 0.38941834 0.43496553 0.47942554 0.52268723
 0.56464247 0.60518641 0.64421769 0.68163876 0.71735609 0.75128041
 0.78332691 0.8134155  0.84147098]


array([1.        , 0.99875026, 0.99500417, 0.98877108, 0.98006658,
       0.96891242, 0.95533649, 0.93937271, 0.92106099, 0.9004471 ,
       0.87758256, 0.85252452, 0.82533561, 0.7960838 , 0.76484219,
       0.73168887, 0.69670671, 0.65998315, 0.62160997, 0.58168309,
       0.54030231, 0.        , 0.04997917, 0.09983342, 0.14943813,
       0.19866933, 0.24740396, 0.29552021, 0.34289781, 0.38941834,
       0.43496553, 0.47942554, 0.52268723, 0.56464247, 0.60518641,
       0.64421769, 0.68163876, 0.71735609, 0.75128041, 0.78332691,
       0.8134155 , 0.84147098])

9. Создайте матрицу $A$ размера $3×5$ из случайных чисел с равномерным распределением на отрезке $[-1,3]$ (используйте np.random.rand)

In [32]:
r = (3 - (-1)) * np.random.random_sample((3, 5)) + (-1)
r

array([[ 2.50738501, -0.54579773, -0.31441231,  0.66849078,  2.13697829],
       [ 0.64009812,  1.41201517,  1.70204924, -0.97875575,  0.32096222],
       [ 0.55764275,  0.37499668, -0.32731056,  0.32077692,  2.22554275]])

10. Найдите сумму всех элементов, сумму внутри строк, сумму внутри столбцов, а также среднее значение, дисперсию и стандартное отклонение чисел для каждой строки матрицы $A$. (Подобно тому, как sorted имеет необязательный аргумент key=, многие функции Numpy имеют необязательный аргумент axis=)


In [33]:
A = np.arange(5,25).reshape(4, 5)
sum = np.sum(A)
sum_row = np.sum(A, axis=-1)
sum_col = np.sum(A.transpose(), axis=-1)
aver = np.average(A, axis=-1)
D = np.var(A, axis=-1, dtype=float)
sigma = np.sqrt(D)
print('сумма:', sum)
print('сумма внутри строк:', sum_row)
print('сумма внутри столбцов:', sum_col)
print('среднее значение:', aver)
print('дисперсия:', D)
print('стандартное отклонение:', sigma)

сумма: 290
сумма внутри строк: [ 35  60  85 110]
сумма внутри столбцов: [50 54 58 62 66]
среднее значение: [ 7. 12. 17. 22.]
дисперсия: [2. 2. 2. 2.]
стандартное отклонение: [1.41421356 1.41421356 1.41421356 1.41421356]


11.  Отнимите от каждого элемента матрицы $A$ среднее по строке и поделите на стандартное отклонение строки.

In [34]:
A = np.arange(5,25).reshape(4, 5)
aver = np.average(A, axis=-1)
sigma = np.sqrt(np.var(A, axis=-1, dtype=float))
print((A - aver.reshape(len(aver), 1)) / sigma.reshape(len(sigma), 1))

[[-1.41421356 -0.70710678  0.          0.70710678  1.41421356]
 [-1.41421356 -0.70710678  0.          0.70710678  1.41421356]
 [-1.41421356 -0.70710678  0.          0.70710678  1.41421356]
 [-1.41421356 -0.70710678  0.          0.70710678  1.41421356]]


12. Отсортируйте матрицу $A$ по 3-ому столбцу, т.е. поменяйте местами строки матрицы так, чтобы 3-й столбец оказался отсортированным. Используйте для этого slicing + argsort + indexing.

In [40]:
import numpy as np
A = np.arange(25,5,-1).reshape(4, 5).transpose()
A[A[:, 2].argsort()]

array([[21, 16, 11,  6],
       [22, 17, 12,  7],
       [23, 18, 13,  8],
       [24, 19, 14,  9],
       [25, 20, 15, 10]])

13. Посчитайте две матрицы: $B = A^TA$ и $C = AA^T$. Что вы можете о них сказать? Какого они размера? В чём их сходство и отличие?

In [41]:
import numpy as np
A = np.arange(5,25).reshape(4, 5)
B = A.transpose() @ A
C = A @ A.transpose()
print('Matrix B:\n', B, '\nShape:', B.shape)
print('Matrix C:\n', C, '\nShape:', C.shape)

Matrix B:
 [[ 750  800  850  900  950]
 [ 800  854  908  962 1016]
 [ 850  908  966 1024 1082]
 [ 900  962 1024 1086 1148]
 [ 950 1016 1082 1148 1214]] 
Shape: (5, 5)
Matrix C:
 [[ 255  430  605  780]
 [ 430  730 1030 1330]
 [ 605 1030 1455 1880]
 [ 780 1330 1880 2430]] 
Shape: (4, 4)


Они обе квадратные, но с разным размером


14. Найдите сумму диагональных элементов матриц $B$ и $C$.

In [42]:
A = np.arange(5,25).reshape(4, 5)
B = A.transpose() @ A
C = A @ A.transpose()
sum_B = np.sum(np.diagonal(B))
sum_C = np.sum(np.diagonal(C))
print('Сумма диагональных элементов матрицы B:', sum_B)
print('Сумма диагональных элементов матрицы C:', sum_C)

Сумма диагональных элементов матрицы B: 4870
Сумма диагональных элементов матрицы C: 4870


15.	Посчитайте детерминант и ранг матриц $A$, $B$ и $C$, посмотрите на спектр (набор собственных значений) матриц $B$ и $C$, какие выводы вы можете сделать?  
(подсказка: используйте готовые функции из библиотеки np.linalg). 


In [44]:
A = np.arange(5,25).reshape(4, 5)
B = A.transpose() @ A
C = A @ A.transpose()
rang_A = np.linalg.matrix_rank(A)
rang_B = np.linalg.matrix_rank(B)
rang_C = np.linalg.matrix_rank(C)

det_A = 0
if A.shape[0] == A.shape[1] == rank_A:
    det_A = np.linalg.det(A)
det_B = np.linalg.det(B)
det_C = np.linalg.det(C)

print('The matrix A:\n', A, '\nранг матрицы A =', rang_A, '\n детерминант A =', det_A, end='\n\n')
print('The matrix B:\n', B, '\nранг матрицы B =', rang_B, '\nдетерминант B =', det_B, end='\n\n')
print('The matrix C:\n', C, '\nранг матрицы C =', rang_C, '\nдетерминант C =', det_C)

The matrix A:
 [[ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]] 
ранг матрицы A = 2 
 детерминант A = 0

The matrix B:
 [[ 750  800  850  900  950]
 [ 800  854  908  962 1016]
 [ 850  908  966 1024 1082]
 [ 900  962 1024 1086 1148]
 [ 950 1016 1082 1148 1214]] 
ранг матрицы B = 2 
детерминант B = 0.0

The matrix C:
 [[ 255  430  605  780]
 [ 430  730 1030 1330]
 [ 605 1030 1455 1880]
 [ 780 1330 1880 2430]] 
ранг матрицы C = 2 
детерминант C = -6.361374339702277e-23


Все матрицы одного ранга


16. Численно посчитать определённый интеграл  3-мя методами:  
a)	Прямоугольников $\int\limits_a^b {f(x)dx \approx \Delta x\sum\limits_{k = 0}^N {f({x_k})} }$  
b)	Трапеций $\int\limits_a^b {f(x)dx \approx \frac{{\Delta x}}{2}\left( {f({x_0}) + 2\sum\limits_{k = 1}^{N - 1} {f({x_k})}  + f({x_N})} \right)}$  
c)	По правилу Симпсона  $\int\limits_a^b {f(x)dx \approx \frac{{\Delta x}}{3}\left( {f({x_0}) + 2\sum\limits_{k = 1}^{N/2 - 1} {f({x_{2k}})}  + 4\sum\limits_{k = 1}^{N/2} {f({x_{2k - 1}})}  + f({x_N})} \right)}$  


В реализации этих методов цикл for для суммирования использовать нельзя.  
Нужно посчитать значение интеграла с некоторым шагом интегрирования (например начать с 0.1), потом уменьшить его в десять раз и ещё раз посчитать и так до тех пор пока отличие в ответах будет в 5-м знаке после запятой, т.е. чтобы ошибка была меньше 1e-5. Получить ответы для 3-х разных методов и сделать соответствующие выводы.  

Далее необходимо сравнить полученные оценки с аналитическим решением (т.е. формульным, которое вы должны сами посчитать на листочке и вбить формулу-ответ).   

В качестве интеграла взять один из следующих с номером, сгенерированным случайным образом с помощью функции np.random.randint(10), в качестве seed для генератора случайных чисел взять номер своего студенческого билета.

<table width="80%"
       align="center">
    <tr>
        <td> 0) $\int\limits_3^4 {\frac{{{x^2} + 3}}{{x - 2}}dx}$ </td>
        <td> 1) $\int\limits_{ - 2}^{ - 1} {\frac{{x + 1}}{{{x^3} - {x^2}}}dx}$ </td>
        <td> 2) $\int\limits_1^2 {\frac{{{e^{1/{x^2}}}}}{{{x^3}}}dx}$ </td>
        <td> 3) $\int\limits_1^e {\frac{{\cos (\ln x)}}{x}dx}$ </td>
        <td> 4) $\int\limits_1^e {\frac{{dx}}{{x(1 + {{\ln }^2}x)}}}$ </td>
    </tr>                
    <tr>
        <td> 5) $\int\limits_0^{\pi /2} {{{\cos }^3}\alpha \;d\alpha }$ </td>
        <td> 6) $\int\limits_0^{1/3} {c{h^2}3x\;dx}$ </td>
        <td> 7) $\int\limits_2^3 {\frac{{dy}}{{{y^2} - 2y - 8}}dy}$ </td>
        <td> 8) $\int\limits_{3/4}^2 {\frac{{dx}}{{\sqrt {2 + 3x - 2{x^2}} }}}$ </td>
        <td> 9) $\int\limits_0^2 {\frac{{2x - 1}}{{2x + 1}}dx}$ </td>
        </tr>
</table>

In [47]:
np.random.seed(1032192934)
n = np.random.randint(10)
print('Variant:', n)

Variant: 4


In [52]:
def f(x):
    return (1)/ (x * (1 + np.log(x)*np.log(x)) )

def int_rectangle_left(f, a, b, step = 0.1):
    Z = np.linspace(a, b, int((b - a) / step) + 1)
    return step * np.sum(f(Z)[:len(Z) - 1])
    
def int_rectangle_right(f, a, b, step = 0.1):
    Z = np.linspace(a, b, int((b - a) / step) + 1)
    return step * np.sum(f(Z)[1:])

def int_trapezoid(f, a, b, step = 0.1):
    Z = np.linspace(a, b, int((b - a) / step) + 1)
    f = f(Z)
    return step / 2 * (f[0] + 2 * np.sum(f[1 : len(Z) - 1]) + f[-1])

def int_Simpson(f, a, b, step = 0.1):
    Z = np.linspace(a, b, 2 * int((b - a) / step) + 1)
    N = len(Z) - 1
    f = f(Z)
    return step / 6 * (f[0] + 2 * np.sum(f[2 : N : 2])  + 4 * np.sum(f[1 : N : 2]) + f[N])
    
def calc_int(int_method, f, a, b, start_step = 0.1, prec = 0.00001):
    step = start_step
    res = int_method(f, a, b, step)
    c = 0
    while abs(res - int_method(f, a, b, step / 10)) >= prec:
        step /= 10
        res = int_method(f, a, b, step)
        c += 1
    return res, c
    
if __name__ == '__main__':
    print(calc_int(int_rectangle_left, f, 1, np.exp(1)))
    print(calc_int(int_rectangle_right, f, 1, np.exp(1)))
    print(calc_int(int_trapezoid, f, 1, np.exp(1)))
    print(calc_int(int_Simpson, f, 1, np.exp(1)))

(0.7854015646987259, 3)
(0.7853932473449123, 4)
(0.7853973276463095, 4)
(0.785397327639104, 4)


Я посчитала этот интеграл вручную и получила $\int\limits_1^e {\frac{{dx}}{{x(1 + {{\ln }^2}x)}}}$ = pi/4 = 0.7853981633974483

Метод прямоугольника по левой стороне: I =0.7854015646987259, n = 3

Метод прямоугольника по правой стороне: I = 0.7853932473449123, n = 4

Метод трапеции: I = 0.7853973276463095, n = 4

Метод симпсона: I = 0.785397327639104, n = 4

I - значение определенного итеграла, n - количество итераций для достижения указанной чтоности при заданных параметрах

Значения не отличаются до четвертого знака после запятой (включительно). Однако, самый точный результат дал метод трапеции.